packages/block-editor/src/components/block-tools/use-block-toolbar-popover-props.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+
+const TOOLBAR_HEIGHT = 72;
+const DEFAULT_PROPS = { __unstableForcePosition: true, __unstableShift: true };
+const RESTRICTED_HEIGHT_PROPS = {
+	__unstableForcePosition: false,
+	__unstableShift: false,
+};
+
+function getProps( contentElement, selectedBlockElement ) {
+	if ( ! contentElement || ! selectedBlockElement ) {
+		return DEFAULT_PROPS;
+	}
+
+	const blockRect = selectedBlockElement.getBoundingClientRect();
+	const contentRect = contentElement.getBoundingClientRect();
+
+	if ( blockRect.top - contentRect.top > TOOLBAR_HEIGHT ) {
+		return DEFAULT_PROPS;
+	}
+
+	// When there's not enough space at the top of the canvas for the toolbar,
+	// enable flipping and disable shifting.
+	return RESTRICTED_HEIGHT_PROPS;
+}
+
+/**
+ * Determines the desired popover positioning behavior, returning a set of appropriate props.
+ *
+ * @param {Object}  elements
+ * @param {Element} elements.contentElement       The DOM element that represents the editor content or canvas.
+ * @param {Element} elements.selectedBlockElement The DOM element that represents the first selected block.
+ *
+ * @return {Object} The popover props used to determine the position of the toolbar.
+ */
+export default function useBlockToolbarPopoverProps( {
+	contentElement,
+	selectedBlockElement,
+} ) {
+	const [ props, setProps ] = useState(
+		getProps( contentElement, selectedBlockElement )
+	);
+
+	useEffect( () => {
```

Comment:
Would using `useLayoutEffect` make any difference here?

Reply:
I don't seem to notice any difference. My take on it is that because the hook triggers a re-render, `useLayoutEffect` isn't needed, but I may be wrong.

Reply:
I believe it could be useful to `useLayoutEffect` because, compared to `useEffect`, it's guaranteed to run after React has performed DOM mutations.

Since we're performing measurements (`updateProps` causes some calls to `getBoundingClientRect`), that's why I thought it would be beneficial to use `useLayoutEffect` compared to `useEffect` in this hook.

What do you think?

Reply:
The part that I was unclear about is what happens if you update state from `useLayoutEffect` vs `useEffect`.

This blog post does seem to indicate that updating state will trigger a synchronous re-render:
https://kentcdodds.com/blog/useeffect-vs-uselayouteffect#uselayouteffect

> This can be useful if you need to make DOM measurements (like getting the scroll position or other styles for an element) and then make DOM mutations or trigger a synchronous re-render by updating state.

And I guess `useEffect` only triggers an asynchronous re-render.

So I think you're right that it would be beneficial. I've updated the code 👍